### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
WIP

### CODE REVIEW

File: src/FSharpVSPowerTools.Core/Utils.fs [hunk 1]
```diff
@@ -29,9 +29,15 @@ module Array =
 
     /// Returns true if one array has trailing elements equal to another's.
     let endsWith (suffix: _ []) (whole: _ []) =
-        whole 
-        |> Array.rev
-        |> startsWith (Array.rev suffix)
+        let rec loop step =
+            if step > suffix.Length then true
+            elif step > whole.Length then false
+            elif suffix.[suffix.Length - step] = whole.[whole.Length - step] then loop (step + 1)
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/Utils.fs]: It's small thing but you can hoist `suffix.Length` and `whole.Length` to values outside the `loop` function.

File: src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs [hunk 2]
```diff
@@ -132,7 +132,7 @@ type SyntaxConstructClassifier
                 let tooltip =
                     vsLanguageService.GetOpenDeclarationTooltip(
                         line, endCol, lineStr, Array.toList idents, project, textDocument.FilePath, source)
-                    |> Async.RunSynchronously
+                    |> Async.RunSynchronously // todo remove RunSynchronous !!!!!!!!!!!!!!!!!!!!!
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs]: I would love to see this 'todo' happens!!!

**Comment by vasily-kirichenko** [comment]
When I profiled, nothing VFPT / FCS related was executed in UI thread :( It must be blocked waiting for a background thread to complete ==> UI freeze.

**Comment by vasily-kirichenko** [comment]
Done :)

**Comment by vasily-kirichenko** [comment]
Although it does not freeze in my scenario anymore, code editing is not absolutely smooth (as in Xamarin, for example).

**Comment by dungpa** [comment]
I doubt we will ever make unused (open) declarations fast enough. Moreover, FSharpBinding in Xamarin Studio didn't have *incremental* multi-project checking last time I checked.

**Comment by vasily-kirichenko** [comment]
Scenario:

* Turn on "Resolve unopened namespaces" and both "Gray out..." (I have not tried with these options off though) 
* Add `Seq.filter (fun x -> Thread )` into this pipe https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/QuickInfoMargin.fs#L119
* Wait until "add open declaration" smart tag appears 
* Play with `Thread` word a bit (delete it, type it again and so on)
* After 30 seconds of this game UI frizzed for very long time, now the lags is about 1 second or less


In short, just try to edit any error-squizzled code (preferably a large area of such code) in a not-so-small project.
